### PR TITLE
Special-case `NumberType` for backwards compatibilty

### PR DESCRIPTION
**Pull request by ansley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61533**

Differential Revision: [D29661173](https://our.internmc.facebook.com/intern/diff/D29661173)

### CODE REVIEW

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29661173).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29661173).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29661173).

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29661173).

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 1]
```diff
@@ -1410,11 +1414,13 @@ struct to_ir {
...
-        std::stringstream err;
-
+        // Special-case annotations that are a subset of NumberType
+        // for backwards compatibility, e.g. `x: Dict[int, T]` should
+        // be a valid lhs for an rhs key of type `number`
         bool is_key_subtype =
-            k->type()->isSubtypeOfExt(dict_type_hint->getKeyType(), &ss);
+            k->type()->isSubtypeOfExt(dict_type_hint->getKeyType(), &ss) ||
+            (k->type() == NumberType::get() &&
+             dict_type_hint->getKeyType()->isSubtypeOfExt(k->type(), &ss));
```
Comment (tugsbayasgalan) [inline torch/csrc/jit/frontend/ir_emitter.cpp:1423]: Just making sure. Are we saying int is a subtype of Number?

Comment (tugsbayasgalan) [inline torch/csrc/jit/frontend/ir_emitter.cpp:1423]: If the Number type is actually not an integer, wouldn't it fail?

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 2]
```diff
@@ -3899,11 +3912,12 @@ struct to_ir {
                 type_hint->expect<DictType>()->getValueType();
             for (size_t i = 0; i < types.size(); ++i) {
               TORCH_CHECK(
-                  types[i]->isSubtypeOf(value_type_hint),
+                  types[i]->isSubtypeOf(value_type_hint) ||
+                      types[i] == NumberType::get(),
                   "Type "
-                  "hint for dict was",
+                  "hint for dict was ",
```
Comment (tugsbayasgalan) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3918]: #grammar :)

**Review by tugsbayasgalan — APPROVED** [review]

**Comment by gmagogsfm** [comment]
I think based on our offline discussion, this is no longer needed because the broken test has been fixed. Closing this PR.

**Pull request closed without merging**